core/shared/src/main/scala/fs2/async/immutable/Signal.scala

```diff
@@ -36,23 +37,91 @@ abstract class Signal[F[_], A] { self =>
   def get: F[A]
 }
 
-object Signal {
+object Signal extends SignalLowPriorityImplicits {
+
+  implicit def applicativeInstance[F[_]](
+      implicit effectEv: Effect[F],
+      ec: ExecutionContext
+  ): Applicative[Signal[F, ?]] =
+    new Applicative[Signal[F, ?]] {
+      override def map[A, B](fa: Signal[F, A])(f: A => B): Signal[F, B] = Signal.map(fa)(f)
+
+      override def pure[A](x: A): Signal[F, A] = fs2.async.mutable.Signal.constant(x)
+
+      override def ap[A, B](ff: Signal[F, A => B])(fa: Signal[F, A]): Signal[F, B] =
```

Comment:
I am a bit curious on `ff`. Eventhough I understand the need for this, I am sort of curious how to define it in reality. You will need for discrete signal (#55) to assure that new value is produced for each ff, and as such gurantee that ff1, ff2, ff3 are different instances. I am not saying this is impossible, but I am feeling a bit awkward here.

Reply:
I don't think I understand what you mean by discrete signal here. I think you could construct one with `Stream.unfold` and `async.hold`?

I agree that `ap` is probably not what most people would use. In fact I'm considering overriding `Applicative.product` and then implementing `ap` in terms of `product` and `map`. The intent here is that they would use `Applicative.product` or `Applicative.mapN` far more. E.g.

```scala
val someSignal: Signal[F, String] = ???
val otherSignal: Signal[F, String] = ???
(someSignal, otherSignal).mapN{ case (x, y)  => (x ++ y).size}
```

Reply:
It might make sense for me to just add a bunch more docs as well to the micro-site, but I don't know if I want to do that here or in a separate PR

Reply:
My trouble is that function always persists its identtity, so : 
```scala 
val  f1: Int => String = (i) => i.toString
val  f2: Int => String = (i) => i.toString
assert(f1 == f1) 
assert(f1 != f2)
assert(f2 == f2)
```
Now imagine signal of functions you need to produce. Essentially you will need to create always new function instance something like :
```
val signal: Signal[Int] = ???

signal.map { i0 => val f1: Int => String = (i) => (i + i0).toString }, to produce signal, that will produce reliable discrete source of Int => String function.  
```
As I said, I am not saying it is impossible, I just hardly believe this is something that users will do and is not perfect code. 

In fact in 0.8 we had similar encodings for sinks and pipes, and we moved from it for several reasons, including performance bottleneck and resource safety.

Reply:
I don't define `ap` here with the expectation that users will use it. I define it purely because that's what is needed for an `Applicative` which I _do_ want because it gives me access to `mapN` and `product` which I am far more likely to use.

`Stream` actually is in the exact same scenario. This compiles:

```scala
import cats.implicits._

Stream((i: Int) => i + 1).ap(Stream(0)) // Stream(1)
```

because `Stream` is `Applicative`, but I would be surprised if anyone ever used `ap`. A similar situation arises here. I define `ap` because that's what makes `Signal` applicative, but I don't think anyone would ever use it directly.

So I agree that `ap` is hard to use, but I don't see users using it. I see them using other interesting parts of the `Applicative` instance. Instead, it makes things like this possible:

```
val s0: Signal[IO, Int] = ???
val s1: Signal[IO, Int] = ???
List(s0, s1).sequence // This gives you a Signal[IO, List[Int]]
```

Reply:
I see, and agree. The reason why I am bringing it up is the fact that 
```
  nondeterministicZip(ff.discrete, fa.discrete).map { case (f, a) => f(a) }
``` 

assumes that `ff.discrete` will change regularly, and actually it will likely only emit single value and then holds.  Not sure of the impacts there, I mean what combinators for Applicative will this broke. 

I hope that all Laws fro applicative covers that and we are safe there.